lib/charms/grafana_k8s/v0/grafana_source.py

```diff
@@ -0,0 +1,433 @@
+                this is `grafana-source`. The values of the dictionary
+                are corresponding minimal acceptable semantic versions
+                for the service.
+            multi: an optional (default `False`) flag to indicate if
+                this object should support interacting with multiple
+                service providers.
+
+        """
+        super().__init__(charm, name, consumes, multi)
+
+        self.charm = charm
+
+        self._stored.set_default(sources=dict())  # available data sources
+        self._stored.set_default(sources_to_delete=set())
+        events = self.charm.on[name]
+        self.framework.observe(events.relation_joined, self._set_sources)
+
+    def add_source(self, data: Dict, rel_id=None) -> None:
+        """Add an additional source to the Grafana source service.
+
+        Args:
+            data: a :dict: object of the source to monitor,
+                in the format of:
+                {
+                "source-type": <source-type>,
+                "address": <address>,
+                "port": <port>,
+                "source-name": <source-name> (optional),
+                }
+            rel_id: an optional integer specifying the relation ID
+                for the grafana source service, only required if the
+                :class:`GrafanaSourceConsumer` has been instantiated in
+                `multi` mode.
+
+        """
+        rel = self.framework.model.get_relation(self.name, rel_id)
+        rel_id = rel_id if rel_id is not None else rel.id
+
+        source_data = SourceData(
+            self.charm.app.name,
+            self.charm.unit,
+            rel.app,
+            rel_id,
+            data,
+        )
+
+        try:
+            data = _validate(self, source_data)
+        except SourceFieldsMissingError as e:
+            logger.critical(
+                f"Missing data on added grafana-k8s source {e}", exc_info=True
+            )
+            return
+        self._update_sources(data, rel_id)
+
+    def remove_source(self, rel_id=None) -> None:
+        """Removes a source relation.
+
+        Args:
+            rel_id: an optional integer specifying the relation ID
+                for the grafana-k8s source service, only required if the
+                :class:`GrafanaSourceConsumer` has been instantiated in
+                `multi` mode.
+        """
+        rel = self.framework.model.get_relation(self.name, rel_id)
+
+        if rel_id is None:
+            rel_id = rel.id
+
+        source = self._stored.sources.pop(rel_id)
+
+        if source is not None:
+            self._stored.sources_to_delete.add(source["source-name"])
+
+    def list_sources(self) -> List[dict]:
+        """Returns an array of currently valid sources"""
+        sources = []
+        for source in self._stored.sources.values():
+            sources.append(source)
+
+        return sources
+
+    @property
+    def removed_source_names(self) -> List[str]:
```

Comment:
It is not obvious how and why this method would be used by the Consumer charm ?

Reply:
Grafana's method of removal is exceptionally stupid. See https://grafana.com/docs/grafana/latest/administration/provisioning/ and look for "deleteDatasources". Any consumer which may want to programatically update their own list or check to see if it's there (Grafana Dashboard being a good example, but there are others) may want to know